### ISSUE TITLE
TypeDoesNotContainType - False positive on isset.

### ISSUE THREAD

**Issue by iaicam** [vimeo/psalm#4088]
The following code will trigger a TypeDoesNotContainType expecting null as type. 

```php
class Random {
 
     private string $type;

    public function getType() {
       if (!isset($this->type)) {
             throw new Exception("Type isn't set");
       }
      return $this->type;
   }
}
```

This is a false positive because it claims that the type has to be nullable to go into `isset`, however it does not. $type has not been set the only way to check it has been set is to call `isset`. I feel like making it nullable removes the benefit I get from having it strictly typed. It can only be a string, if it's anything else including not set or null I should get an error.

**Comment by iaicam** [comment]
Yes.

**Comment by weirdan** [comment]
> Yes.

https://psalm.dev/r/ec9c21b72f

Psalm assumes you would initialize everything in the constructor.

**Comment by muglug** [comment]
This is with the correct typing: https://psalm.dev/r/0d75f88675

**Comment by iaicam** [comment]
I would say the two rules are different. One is a rule to always define things within the constructor. The other is a type rule that says that to be used with isset the variable must be nullable. This is not true, the documentation and behaviour of PHP do not have that to be the case. I would say this rule as is, makes code unsafer and not safer.

https://psalm.dev/r/3008f80d86 - Issue is still there even if variable is set via constructor.
